### PR TITLE
Handle repeated mount point in FsInfo. 

### LINKED ISSUE #27174: Wrong disk free space reported when there are multiple data paths
Moving https://github.com/elastic/kibana/issues/14442 to Elasticsearch repo

Quoting from original author:

> **Kibana version**:
> 5.5.0
> **Elasticsearch version**:
> 5.5.0
> **Server OS version**:
> centos 6.x
> **Browser version**:
> chrom 61.0.3163.91
> **Browser OS version**:
> win 7 64 bit
> **Original install method (e.g. download page, yum, from source, etc.)**:
> download from elastic.co
> **Description of the problem including expected versus actual behavior**:

> I set two path as elasticsearch store data path
> "path" : "/data1/esdata/cc_elastic_node_data1/nodes/0",
> "mount" : "/data1 (/dev/sdb1)",

> "path" : "/data1/esdata/cc_elastic_node_data2/nodes/0",
> "mount" : "/data1 (/dev/sdb1)",

> and my server disk free space (dev/sdb1) is 1.5T
 
> but in kibana monitoring（x-pack）-> nodes -> Disk Free Space it display 3.0 TB, it  calculated same disk twice, is it right?

The value shown in Kibana Monitoring comes from Cluster Stats (`GET _cluster/stats`), which confirms that the calculation for free disk space is doubled when the node is configured to have 2 data paths.

cc @shijiebei2009 @leandrojmp

### PR DESCRIPTION
**Pull request by kiawin**
Closes #27174.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java [hunk 1]
```diff
@@ -496,10 +496,11 @@ public DiskUsage getMostDiskEstimate() {
     private Path total() {
         Path res = new Path();
         Set<String> seenDevices = new HashSet<>(paths.length);
+        Set<String> seenMounts = new HashSet<>(paths.length);
         for (Path subPath : paths) {
             if (subPath.path != null) {
-                if (!seenDevices.add(subPath.path)) {
-                    continue; // already added numbers for this device;
+                if (!seenDevices.add(subPath.path) || !seenMounts.add(subPath.mount)) {
```
Comment (DaveCTurner) [inline core/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java]: This doesn't look right to me. If `subPath.path` is null, or if `seenDevices.add()` returned false then we don't call `seenMounts.add()` despite having seen the mount point in question.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Thanks for bringing a test suite for this code into existence. I added a comment that the change doesn't quite work as I'd expect due to short-circuiting etc. Please could you extend the test suite to test for this, and fix as you see fit?

File: core/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java (hunk 1 above)
Comment (kiawin) [reply core/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java]: Good point, let me work on it

Comment (kiawin) [reply core/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java]: Based on your comments, I made the following assumptions:

1. valid path + valid mount point - OK
2. invalid path + valid mount point - OK
3. valid path + invalid mount point - NOT OK
4. invalid path + invalid mount point - NOT OK

Do update me if you think the above is incorrect.

**Comment by kiawin** [comment]
@DaveCTurner I have added several tests and made few assumptions. Can you review them? Thanks :)

File: core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        Arrays.stream(paths).forEach(p -> p.mount = mount);
+        final long timestamp = randomLong();
+        FsInfo fsInfo = new FsInfo(timestamp, null, paths);
+
+        long size = paths[0].getTotal().getBytes();
+        assertThat(timestamp, equalTo(fsInfo.getTimestamp()));
+        assertThat(size, equalTo(fsInfo.getTotal().getTotal().getBytes()));
+    }
+
+    public void testFsInfoSameDevices() {
```
Comment (DaveCTurner) [inline core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java]: This seems to test the case where the same path has a number of distinct mount points. Can this happen? I can't think how.

File: core/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java [hunk 3]
```diff
@@ -496,11 +496,23 @@ public DiskUsage getMostDiskEstimate() {
...
+            } else {
+                seenDevice = true;
+            }
+            if (subPath.mount != null) {
+                seenMount = !seenMounts.add(subPath.mount);
+            } else {
+                seenMount = true;
+            }
+
+            if (seenMount || (seenDevice && seenMount)) {
```
Comment (DaveCTurner) [inline core/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java]: This condition is equivalent to simply `seenMount`. Effectively, `seenDevice` is unused. I think this means something is missing/wrong in the test cases.

File: core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        Arrays.stream(paths).forEach(p -> p.path = path);
+        final long timestamp = randomLong();
+        FsInfo fsInfo = new FsInfo(timestamp, null, paths);
+
+        long size = Arrays.stream(paths).map(n -> n.getTotal().getBytes()).mapToLong(n -> n).sum();
+        assertThat(timestamp, equalTo(fsInfo.getTimestamp()));
+        assertThat(size, equalTo(fsInfo.getTotal().getTotal().getBytes()));
+    }
+
+    public void testFsInfoNullMount() {
```
Comment (DaveCTurner) [inline core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java]: The existing code doesn't filter out paths whose `mount` property is `null`, so I don't think we should start doing so as per this test case.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I left some comments, questions and requests. There is something suspicious about the implementation given which also makes me suspicious about the test suite, but I haven't reviewed the full suite yet. Could you address these comments and I'll see where that takes us?

File: core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+public class FsInfoTests extends ESTestCase {
+
+    public void testFsInfo() {
+        final FsInfo.Path[] paths = randomPaths();
+        final long timestamp = randomLong();
+        FsInfo fsInfo = new FsInfo(timestamp, null, paths);
+
+        long size = Arrays.stream(paths).map(n -> n.getTotal().getBytes()).mapToLong(n -> n).sum();
+        assertThat(timestamp, equalTo(fsInfo.getTimestamp()));
```
[the remark this answers is not in the document]
Comment (kiawin) [reply core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java]: Sure

File: core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    private FsInfo.Path[] randomPaths() {
+        final int totalPaths = randomIntBetween(2, 4);
+        FsInfo.Path[] paths = new FsInfo.Path[totalPaths];
+        for (int i=0; i < totalPaths; i++) {
+            paths[i] = new FsInfo.Path(
+                randomAlphaOfLength(10),
+                randomAlphaOfLength(10),
+                randomIntBetween(1, 1 << 16),
```
[the remark this answers is not in the document]
Comment (kiawin) [reply core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java]: Sure

File: core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java (hunk 2 above)
Comment (kiawin) [reply core/src/test/java/org/elasticsearch/monitor/fs/FsInfoTests.java]: I will remove this

**Comment by martijnvg** [comment]
@kiawin It has been a while, but do still want to work on this and get @DaveCTurner feedback in?

**Comment by javanna** [comment]
@DaveCTurner it's been a while since this PR was opened and first reviewed, could you have another look to see if comments were addressed? If not shall we take this over or ... ?

**Comment by DaveCTurner** [comment]
My comments were acked but there have been no further changes. I don't have an opinion about whether we should take this over - up to you.

**Comment by javanna** [comment]
Closing as comments were not addressed.

**Pull request closed without merging**